src/opt/pre_scheduling_impl.cpp

```diff
@@ -0,0 +1,326 @@
+#include "pre_scheduling_impl.hpp"
+#include <migraphx/iterator_for.hpp>
+#include <migraphx/pass_config.hpp>
+#include <stack>
+namespace migraphx {
+
+// Compute accumulated weights for each node in the DAG. Collect exit nodes
+// and sort them according to accumulated weights.
+//
+void pre_scheduling_impl::compute_weights()
+{
+    int ndx = 0;
+    std::unordered_map<dag_node*, bool> visited;
+    for(auto ins : iterator_for(*p_program))
+    {
+        dag_node& node             = nodes[ndx];
+        std::pair<int, int> weight = weight_func(ins->get_operator());
+        node.weight                = weight.first;
+        node.run_on_cpu            = weight.second;
```

Comment:
What does `run_on_cpu` mean? And why is it that you dont sum this weight?

Reply:
Ah wait it looks like you only compare this number to `0`. What about the value of this number?

Reply:
That mean the instruction is a CPU-only instruction, which will not be assigned streams.

Reply:
Yes so when its `0` it means it can be assigned a stream and non-zero it cannot. But what does the value mean when its non-zero? Whats the difference between `run_on_cpu` set to `1` or set to `10 or `-1`?

Reply:
Non-zero are !run_on_cpu,   zero is run_on_cpu,  I can change it to bool.

Reply:
Ah that makes a lot more sense, it supposed to a bool that can be used to know whether a stream can be assigned. Since its much better to use `is_context_free` for this, we can remove this all together.

Reply:
Does is_context_free check whether "compute" takes context as input?   There are operators that does not have "compute" method at all.

Reply:
> Does is_context_free check whether "compute" takes context as input?

Yes, `is_context_free` returns `true` if the compute method does not take a `context`. Which means the operator will not run on the gpu as the compute method has no access to the stream.

> There are operators that does not have "compute" method at all.

We should return `true` for this case, but I dont think we do. Let me update `is_context_free`. As a workaround, you can implement `can_use_stream` as `return is_context_free(ins->get_operator()) and ins->name().front() == "@";` since the only operators not computable at this point should just be the builtins(which only the builtin names start with `@`).

Reply:
Using is_context_free to identify whether an instruction can run on GPU leads to incorrect stream assignments as shown below.   Apparently, param, hip::load_literal and load are mistakenly considered as GPU instructions.   I also recall why I use "int" type instead of "bool":  a 32-bit int can encode 32 boolean fields.   Passing an integer instead of a structure of many bool fields is cleaner approach.   For this check-in, I left this as it is.   We will use is_context_free when it is fully tested and supported.

0 = @param:0(stream=0) -> float_type, {1, 3, 299, 299}, {268203, 89401, 299, 1}: 0.00223892ms, 1%
@4 = hip::load_literal(stream=0) -> float_type, {32, 3, 3, 3}, {27, 9, 3, 1}: 0.00208692ms, 1%
@12 = load[shape=float_type, {1, 32, 149, 149}, {710432, 22201, 149, 1},offset=5431552](scratch)(stream=0) -> float_type, {1, 32, 149, 149}, {710432, 22201, 149, 1}: 0.00220162ms, 1%

Reply:
> Apparently, param, hip::load_literal and load are mistakenly considered as GPU instructions.

`load_literal` is a GPU instruction, but `param` and `load` is not.

Reply:
What I mean gpu instruction is:  the instruction launches a kernel.  Otherwise, there is no need to assign a stream to it.